src/Microsoft.NetCore.Analyzers/Core/Runtime/InstantiateArgumentExceptionsCorrectly.Fixer.cs

```diff
@@ -11,19 +23,379 @@ namespace Microsoft.NetCore.Analyzers.Runtime
     /// </summary>
     public abstract class InstantiateArgumentExceptionsCorrectlyFixer : CodeFixProvider
     {
-        public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(InstantiateArgumentExceptionsCorrectlyAnalyzer.RuleId);
+        private class SemanticParameterComparer : IEqualityComparer<IParameterSymbol>
+        {
+            public static readonly SemanticParameterComparer Instance = new SemanticParameterComparer();
+
+            private SemanticParameterComparer()
+            {
+            }
+
+            public bool Equals(IParameterSymbol x, IParameterSymbol y)
+            {
+                return x.Name == y.Name && x.Type == y.Type;
+            }
+
+            public int GetHashCode(IParameterSymbol obj)
+            {
+                return unchecked(obj.Name.GetHashCode() * 33 + obj.Type.GetHashCode());
+            }
+        }
+
+        public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(RuleId);
+
+        public sealed override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
+
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var document = context.Document;
+            var root = await document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var semanticModel = await document.GetSemanticModelAsync(context.CancellationToken).ConfigureAwait(false);
+
+            SyntaxNode creationSyntax = root.FindNode(context.Span, getInnermostNodeForTie: true);
+            var diagnostics = context.Diagnostics;
+
+            switch (GetProblemKind(diagnostics[0]))
+            {
+                case ProblemKind.IncorrectMessage:
+                    if (diagnostics.Length == 2 && GetProblemKind(diagnostics[1]) == ProblemKind.IncorrectParameterName)
```

Comment:
@333fred I've managed to come pretty close to finishing this PR (it's been pretty fun, too :smile:). I ran into an issue though. For expressions like

```cs
// message and parameter name are swapped
new ArgumentNullException("value is null", nameof(value))
```

the analyzer should fire twice, since both a parameter name is being passed to `message` and a message is being passed to `paramName`. I've tried to detect such scenarios here by checking `diagnostics.Length == 2`, so in that case I can just swap the args. But I'm only getting a diagnostics array of length 1; do you know why?

Reply:
Are you sure the Span being given here has both parameters in it?

Separately: if they're swapped, I wouldn't expect 2 diagnostics. I'd expect 1, with that as the reason. I'd modify the analyzer to detect that and have a different message.